File: sample/swift/03_julia_set.swift
import Foundation


// 03: Sample that outputs a Julia set as a PNG image.
// Implemented with simple loop-centric logic for transpilation compatibility.

func render_julia(width: Int64, height: Int64, max_iter: Int64, cx: Double, cy: Double) -> [Any] {
    var pixels: [Any] = __pytra_as_list([])
    var __hoisted_cast_1: Double = __pytra_float(__pytra_float((__pytra_int(height) - __pytra_int(Int64(1)))))
    var __hoisted_cast_2: Double = __pytra_float(__pytra_float((__pytra_int(width) - __pytra_int(Int64(1)))))
    var __hoisted_cast_3: Double = __pytra_float(__pytra_float(max_iter))
    let __step_0 = __pytra_int(Int64(1))
    var y = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && y < __pytra_int(height)) || (__step_0 < 0 && y > __pytra_int(height))) {
        var zy0: Double = __pytra_float((__pytra_float((-Double(1.2))) + __pytra_float((__pytra_float(Double(2.4)) * __pytra_float((__pytra_float(y) / __pytra_float(__hoisted_cast_1)))))))
        let __step_1 = __pytra_int(Int64(1))
        var x = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && x < __pytra_int(width)) || (__step_1 < 0 && x > __pytra_int(width))) {
            var zx: Double = __pytra_float((__pytra_float((-Double(1.8))) + __pytra_float((__pytra_float(Double(3.6)) * __pytra_float((__pytra_float(x) / __pytra_float(__hoisted_cast_2)))))))
            var zy: Double = __pytra_float(zy0)
            var i: Int64 = __pytra_int(Int64(0))
            while (__pytra_int(i) < __pytra_int(max_iter)) {
                var zx2: Double = __pytra_float((__pytra_float(zx) * __pytra_float(zx)))
                var zy2: Double = __pytra_float((__pytra_float(zy) * __pytra_float(zy)))
                if (__pytra_float((__pytra_float(zx2) + __pytra_float(zy2))) > __pytra_float(Double(4.0))) {
                    break
                }
                zy = __pytra_float((__pytra_float((__pytra_float((__pytra_float(Double(2.0)) * __pytra_float(zx))) * __pytra_float(zy))) + __pytra_float(cy)))
                zx = __pytra_float((__pytra_float((__pytra_float(zx2) - __pytra_float(zy2))) + __pytra_float(cx)))
                i += Int64(1)
            }
            var r: Int64 = __pytra_int(Int64(0))
            var g: Int64 = __pytra_int(Int64(0))
            var b: Int64 = __pytra_int(Int64(0))
            if (__pytra_int(i) >= __pytra_int(max_iter)) {
                r = __pytra_int(Int64(0))
                g = __pytra_int(Int64(0))
                b = __pytra_int(Int64(0))
            } else {
                var t: Double = __pytra_float((__pytra_float(i) / __pytra_float(__hoisted_cast_3)))
                r = __pytra_int(__pytra_int((__pytra_float(Double(255.0)) * __pytra_float((__pytra_float(Double(0.2)) + __pytra_float((__pytra_float(Double(0.8)) * __pytra_float(t))))))))
                g = __pytra_int(__pytra_int((__pytra_float(Double(255.0)) * __pytra_float((__pytra_float(Double(0.1)) + __pytra_float((__pytra_float(Double(0.9)) * __pytra_float((__pytra_float(t) * __pytra_float(t))))))))))
                b = __pytra_int(__pytra_int((__pytra_float(Double(255.0)) * __pytra_float((__pytra_float(Double(1.0)) - __pytra_float(t))))))
            }
            pixels = __pytra_as_list(pixels); pixels.append(r)
            pixels = __pytra_as_list(pixels); pixels.append(g)
            pixels = __pytra_as_list(pixels); pixels.append(b)
            x += __step_1
        }
        y += __step_0
    }
    return pixels
}

func run_julia() {
    var width: Int64 = __pytra_int(Int64(3840))
    var height: Int64 = __pytra_int(Int64(2160))
    var max_iter: Int64 = __pytra_int(Int64(20000))
    var out_path: String = __pytra_str("sample/out/03_julia_set.png")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var pixels: [Any] = __pytra_as_list(render_julia(width, height, max_iter, (-Double(0.8)), Double(0.156)))
    __pytra_noop(out_path, width, height, pixels)
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("size:", width, "x", height)
    __pytra_print("max_iter:", max_iter)
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_julia()
    }
}

File: sample/go/03_julia_set.go
package main


// 03: Sample that outputs a Julia set as a PNG image.
// Implemented with simple loop-centric logic for transpilation compatibility.

func render_julia(width int64, height int64, max_iter int64, cx float64, cy float64) []any {
    var pixels []any = __pytra_as_list([]any{})
    var __hoisted_cast_1 float64 = __pytra_float((height - int64(1)))
    var __hoisted_cast_2 float64 = __pytra_float((width - int64(1)))
    var __hoisted_cast_3 float64 = __pytra_float(max_iter)
    for y := int64(0); y < height; y += 1 {
        var zy0 float64 = ((-float64(1.2)) + (float64(2.4) * (float64(y) / __hoisted_cast_1)))
        for x := int64(0); x < width; x += 1 {
            var zx float64 = ((-float64(1.8)) + (float64(3.6) * (float64(x) / __hoisted_cast_2)))
            var zy float64 = zy0
            var i int64 = int64(0)
            for (i < max_iter) {
                var zx2 float64 = (zx * zx)
                var zy2 float64 = (zy * zy)
                if ((zx2 + zy2) > float64(4.0)) {
                    break
                }
                zy = (((float64(2.0) * zx) * zy) + cy)
                zx = ((zx2 - zy2) + cx)
                i += int64(1)
            }
            var r int64 = int64(0)
            var g int64 = int64(0)
            var b int64 = int64(0)
            if (i >= max_iter) {
                r = int64(0)
                g = int64(0)
                b = int64(0)
            } else {
                var t float64 = (float64(i) / __hoisted_cast_3)
                r = __pytra_int((float64(255.0) * (float64(0.2) + (float64(0.8) * t))))
                g = __pytra_int((float64(255.0) * (float64(0.1) + (float64(0.9) * (t * t)))))
                b = __pytra_int((float64(255.0) * (float64(1.0) - t)))
            }
            pixels = append(pixels, r)
            pixels = append(pixels, g)
            pixels = append(pixels, b)
        }
    }
    return __pytra_as_list(pixels)
}

func run_julia() {
    var width int64 = int64(3840)
    var height int64 = int64(2160)
    var max_iter int64 = int64(20000)
    var out_path string = __pytra_str("sample/out/03_julia_set.png")
    var start float64 = __pytra_perf_counter()
    var pixels []any = __pytra_as_list(render_julia(width, height, max_iter, (-float64(0.8)), float64(0.156)))
    __pytra_write_rgb_png(out_path, width, height, pixels)
    var elapsed float64 = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("size:", width, "x", height)
    __pytra_print("max_iter:", max_iter)
    __pytra_print("elapsed_sec:", elapsed)
}

func main() {
    run_julia()
}
